---
title: "GWAS_models"
author: "Filippo Biscarini"
date: "October 19, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("ggplot2")
library("knitr")
```

## GWAS: quantitative traits

For quantitative traits (continuous target variables), the basic model is a simple linear regression model of the following form:

$$
y_{ij}=\mu+SNP_j+e_{ij}
$$

where $y_{ij}$ is the phenotypic observation for sample *i* with SNP genotype *j*, $SNP_j$ is the effect of SNP genotype *j* (e.g. AA, AB, BB), and $e_{ij}$ are the residuals of the model.

```{r data}
setwd("~/Dropbox/cursos/berlin2018/data/")
phenotypes <- read.table("phenotypes_rice.txt", header = TRUE)
genotypes <- read.table("rice_snp.raw", header = TRUE)
genotypes <- genotypes[,-c(1,3,4,5,6)]

phenotypes <- merge(phenotypes, genotypes, by.x="id", by.y="IID")
names(phenotypes) <- c("id","population","PH","sex","snp")

kable(head(phenotypes))
```

### Modelling number of copies of the minor allele

First, we can model SNP genotypes as the number of copies of the minor allele: 0, 1, 2 (numeric variable)

```{r model1}
fit <- lm(PH~snp, data = phenotypes)
```

```{r}
g <- summary(fit)
g
```

#### ANOVA table

```{r}
kable(anova(fit))
```

### Modelling genotypes

Genotypes can be coded as AA, AB and BB (factor variables), instead. The p-value is not affected, but the estimated SNP effect is.

```{r model2}
phenotypes$snp <- as.factor(phenotypes$snp)
fit <- lm(PH~snp, data = phenotypes)
```

```{r}
g <- summary(fit)
g
```

```{r, echo=FALSE}
kable(anova(fit))
```

## Adding population structure

```{r}
fit <- lm(PH~population+snp, data = phenotypes)
```

```{r, echo=FALSE}
g <- summary(fit)
g
```

```{r}
anova(fit)
```

## GWAS: binary traits

For binary traits (e.g. case/control), the basic model is a simple logistic regression model of the following form:

$$
\left\{ 
\begin{array}{lr}
P(y_i=1|x_i)=p(x_i) \\
\text{logit}(p(x_i))=\text{log} \left( \frac{p(x_i)}{1-p(x_i)} \right)=\mu+\sum_{j=1}^m z_{ij} SNP_j
\end{array}
\right.
$$

where $P(y_i=1|x_i)$ is the probability for sample *i* to belong to group "1" given its SNP genotypes *x_i*.

```{r data_sheep, echo=FALSE}
setwd("~/Dropbox/cursos/berlin2018/data/")
phenotypes <- read.table("pheno_genabel.dat", header = TRUE)
genotypes <- read.table("sheep_snp.raw", header = TRUE)
genotypes <- genotypes[,-c(1,3,4,5,6)]

phenotypes <- merge(phenotypes, genotypes, by.x="id", by.y="IID")
names(phenotypes) <- c("id","breed","horns","sex","snp")

kable(head(phenotypes))
```

Prepare phenotype data:

```{r}
phenotypes <- phenotypes[phenotypes$horns!=0,]
phenotypes$horns <- phenotypes$horns-1
```

### Fit a generalised linear model

We now need to fit a generalised linear model instead of a simple linear regression model: since we are analysing binary data, a common choice is to use the **binomial distribution** and the **logit()** link function: 

```{r}
fit <- glm(horns~snp, data = phenotypes, family=binomial(link="logit"))
```

```{r, echo=FALSE}
g <- summary(fit)
g
```

```{r, echo=FALSE}
full_model <- glm(horns~snp, data = phenotypes, family="binomial")
reduced_model <- glm(horns~1, data = phenotypes, family="binomial")
anova(reduced_model, full_model, test = "LRT")

l1 <- logLik(full_model)
l0 <- logLik(reduced_model)
df <- length(coef(full_model)) - length(coef(reduced_model)) 

teststat<--2*(as.numeric(l0)-as.numeric(l1))
pchisq(teststat,df=1,lower.tail=FALSE)
```


