---
title: "GWAS - binary trait"
author: "Filippo Biscarini"
date: "October 10, 2018"
output: html_document
bibliography: gwas.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("knitr")
library("tidyr")
library("dplyr")
library("GenABEL")
library("ggplot2")
library("reshape2")
```

## GWAS for a binary trait

From Kijas et al. @kijas2016genome, data on the number of horns (two/four) in sheep have been obtained.


```{r data, echo=FALSE}
setwd("~/Dropbox/cursos/berlin2018")
load("data/df_genabel.RData")
```

### Description of marker data

GenABEL provides functions to look in a compact way to the main features of marker data:

#### MAF

```{r, echo=FALSE}
D <- descriptives.marker(df)
kable(D$`Minor allele frequency distribution`)
```

#### HWE

```{r, echo=FALSE}
kable(D$`Cumulative distr. of number of SNPs out of HWE, at different alpha`)
```

#### call-rate per sample

```{r, echo=FALSE}
kable(D$`Distribution of proportion of successful genotypes (per person)`)
```

#### call-rate per SNP

```{r, echo=FALSE}
kable(D$`Distribution of proportion of successful genotypes (per SNP)`)
```

#### Distribution of markers per chromosome

N. `r gtdata(df)@nsnps` from `r length(unique(gtdata(df)@chromosome))` chromosomes

```{r, echo=FALSE}
dd <- as.data.frame(table(gtdata(df)@chromosome))
names(dd) <- c("chrom","nSnps")
kable(dd)
```

### Description of phenotypes

Two breeds, three phenotypic categories:

- **1** = two horns
- **2** = four horns
- **0** = unknown

```{r, echo=FALSE}
kable(descriptives.trait(df))

phdata(df) %>%
  group_by(breed,phenotype) %>%
  summarise(N=n()) %>%
  spread(key=breed, value=N)
  
```

## Barplot of phenotypic categories

For the moment, our "phenotype" coincides with the breed:

```{r pressure, echo=FALSE}
pp <- phdata(df)
pp$phenotype <- as.factor(pp$phenotype)
p <- ggplot(pp,aes(x=phenotype)) + geom_bar(aes(fill=phenotype))
p <- p + facet_wrap(~breed) + xlab("phenotype")
print(p)
```

### Quality check

GenABEL provides the function *check.marker()* to perform quality checks (MAF, call-rate, HWE, etc.) on marker data. Most of these have been already performed earlier using Plink. However, we can play around with some of the options and values to see what happens and how sensitive our GWAS model is to different qulity filters.

```{r, label='quality_check', echo=FALSE}
qc1 <- check.marker(df, p.level=0, maf=0.01)
df1 <- df[,qc1$snpok]
```

We now remove samples with unknown phenotypic status (phenotype == 0). This way, our phenotype becomes binary (0/1)

```{r}
## phenotypes: exclude unknown status
df2 <- df1[phdata(df1)$phenotype!=0,]
```

### Kinship matrix

```{r, label='kinship', echo=FALSE}
K <- ibs(df2,weight = "freq")
K[upper.tri(K)] = t(K)[upper.tri(K)]
heatmap(K,col=rev(heat.colors(75)))
```

```{r, label='polygenic'}
ff <- phdata(df2)$phenotype
ff[ff==2] <- 0
df2 <- add.phdata(df2,ff, name = "pheno_recoded")

# h2a <- polygenic(pheno_recoded,data=df2,kin=K,trait.type = "binomial",fixh2 = 0.5,llfun = "polylik")
# df.mm <- mmscore(h2a,df1)
# descriptives.scan(df.mm,top=100)
# plot(df.mm,col = c("red", "slateblue"),pch = 19, cex = .5, main="phenotype")
```

## Run the GWAS model

GWAS without considering population structure (we have two sheep breeds):

```{r, echo=FALSE}
data2.qt <- qtscore(pheno_recoded, data = df2, trait="binomial")
lambda(data2.qt)
plot(data2.qt, df="Pc1df",col = c("red", "slateblue"),pch = 19, cex = .5, main="trait")
descriptives.scan(data2.qt,top=10)
```

We run our GWAS model correcting for population structure

```{r, label='polygenic', results='hide'}
h2a <- polygenic(pheno_recoded,data=df2,kin=K,trait.type = "binomial")
df.mm <- mmscore(h2a,df1)
descriptives.scan(df.mm,top=10)
plot(df.mm,col = c("red", "slateblue"),pch = 19, cex = .5, main="phenotype")
```

```{r, label='eigenn_decomposition', echo=FALSE}
df.mm  <- egscore(phdata(df2)$phenotype,data=df2,kin=K)
lambda(df.mm)
plot(df.mm,col = c("red", "slateblue"),pch = 19, cex = .5, main="trait")
descriptives.scan(df.mm,top=10)
```


## References
